vmware_guest: split in 3 modules (guest, guest_facts & guest_snapshot)

nerzhul opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request
 - New Module Pull Request

##### COMPONENT NAME
vmware_guest
vmware_guest_vm_facts
vmware_guest_snapshot

##### ANSIBLE VERSION
```
ansible 2.3 devel
```

##### SUMMARY

Split vmware_guest in 3 parts: vmware_guest, vmware_guest_facts, vmware_guest_snapshot

New module: vmware_guest_facts

* Move gather_facts action outside of vmware_guest module
* Move facts gathering method to vmware.py library
* vmware_guest use some vmware.py library methods
* vmware.py lib: fix some PEP 8 issues
* facts now returns snapshot list if available

New module: vmware_guest_snapshot

* Create, remove(all), revert snapshots

nerzhul committed:
New modules: vmware_guest_facts & vmware_guest_snapshot

* Move gather_facts action outside of vmware_guest module
* Move facts gathering method to vmware.py library
* vmware_guest use some vmware.py library methods
* vmware.py lib: fix some PEP 8 issues
* Move snapshot actions to vmware_guest_snapshot
* Move common snapshot facts to vmware.py library

nerzhul commented:
@dagwieers i moved facts outside of vmware_guest module while hacking at ansible meetup in paris

cc @gundalow

dagwieers commented:
You rule ! Let's meet !

dagwieers commented:
Why vmware_guest_vm_facts and not vmware_guest_facts.
Are there any non-vm guests in vnware ?

nerzhul commented:
@dagwieers you are right, it's too long, changing it now and fixing changelog.md and looking for shippable if it wants to load on this wifi :(

nerzhul commented:
@dagwieers rebased, force pushed, renamed :)

nerzhul commented:
seems @ansibot is at Ansible Automates event and doesn't want to tag :(

dagwieers reviewed (approved):
Cannot test, but since this is mostly relocating existing code, I am confident this is quite safe.

nerzhul commented:
@dagwieers i added a second commit and moved snapshot part to a third module to manage snapshots only, without the listing parts, which needs more tuning

nerzhul commented:
It's now splited in four parts and need testing @dagwieers when you are ready. I will test all tomorrow more heavily at work

dagwieers commented:
Is there really a need to have 4 different modules ?
Can't we leave the snapshot facts as part of the vmware_guest_facts ?

nerzhul commented:
@dagwieers i can merge the snapshot facts into the VM itself, it's possible yes, if needed. @dav1x @jctanner what do you think about this case ?

dagwieers commented:
Given the fact that so many *_facts modules already exist for other large and small subsystems I don't mind if this get merged as-is. However I hope that all these *_facts modules eventually and up being facts plugins in a new pluggable facts mechanism that we still have to design :-)

jctanner commented:
@nerzhul yeah, let's combine the two facts modules to reduce confusion for the end users.

jctanner reviewed (dismissed):
please combine the facts modules

ritzk commented on lib/ansible/module_utils/vmware.py:
```diff
@@ -181,6 +185,69 @@ def find_host_portgroup_by_name(host, portgroup_name):
     return None
 
 
+def gather_vm_facts(content, vm):
```
Is gather_vm_facts used by any other module ?

dagwieers replied:
Yes it is. (Answering the review)

nerzhul commented:
Yes, it's used by VMware guest when doing actions on a machine and the guest facts

KroArtem commented:
While you're at it, can you please fix this one: https://github.com/ansible/ansible/issues/20070
It's just a ctrl-c, ctrl-v action. :)

nerzhul commented:
WIll do the module merge and re-check the snapshot problem today at work if i'm not too busy

KroArtem commented:
Thanks!

nerzhul commented:
@gundalow @dagwieers @KroArtem @jctanner all concerns are now fixes.

We now have 3 modules. I re-tested and do some fixed on vmware_guest_snapshot, it works perfect on my 5.5 cluster. I tested, add 10 snapshot, revert, re-revert, remove a snapshot in the middle, the first snapshot, remove_all, everything is okay (except my machine which became crazy after many snapshot changes operations :p but after restart it's okay

dagwieers commented:
Looking good to me !
shipit

jctanner commented:
@nerzhul looks like you have a few code style problems ...

```
2017-02-06 10:13:57 Run command: pep8 --max-line-length 160 --config /dev/null --ignore E123,E124,E127,E128,E201,E202,E203,E211,E221,E222,E225,E226,E227,E228,E231,E241,E251,E261,E262,E265,E266,E301,E302,E303,E402,E502,E713,E731,W391,W503 lib/ansible/module_utils/vmware.py lib/ansible/modules/cloud/vmware/vmware_guest.py lib/ansible/modules/cloud/vmware/vmware_guest_facts.py lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py
2017-02-06 10:13:58 ERROR: PEP 8: lib/ansible/module_utils/vmware.py: Passes current rule set. Remove from legacy list (test/sanity/pep8/legacy-files.txt).
2017-02-06 10:13:58 PEP 8: There were 1 different legacy issues found (3 total):
2017-02-06 10:13:58 PEP 8: Count Code Message
2017-02-06 10:13:58 PEP 8:     3 E501 line too long (NNN > NNN characters)
2017-02-06 10:13:58 ERROR: PEP 8: There are 1 issues which need to be resolved.
2017-02-06 10:13:58 ERROR: InvocationError: '/root/src/github.com/ansible/ansible/test/runner/test.py sanity -v --skip-test ansible-doc --python 2.7 --require CHANGELOG.md --require lib/ansible/module_utils/vmware.py --require lib/ansible/modules/cloud/vmware/vmware_guest.py --require lib/ansible/modules/cloud/vmware/vmware_guest_facts.py --require lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py --color yes --requirements'
2017-02-06 10:13:58 ___________________________________ summary ____________________________________
2017-02-06 10:13:58 ERROR:   py27: commands failed
2017-02-06 10:13:58 ERROR: Command "tox -c test/runner/tox.ini -e py27 -- /root/src/github.com/ansible/ansible/test/runner/test.py sanity -v --skip-test ansible-doc --python 2.7 --require CHANGELOG.md --require lib/ansible/module_utils/vmware.py --require lib/ansible/modules/cloud/vmware/vmware_guest.py --require lib/ansible/modules/cloud/vmware/vmware_guest_facts.py --require lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py --color yes --requirements" returned exit status 1.
2017-02-06 10:13:58 + cleanup
2017-02-06 10:13:58 ++ ls test/results/coverage/
2017-02-06 10:13:58 + '[' '' ']'
2017-02-06 10:13:58 + rmdir shippable/testresults/
2017-02-06 10:13:58 + cp -av test/results/junit/ shippable/testresults/
2017-02-06 10:13:58 ‘test/results/junit/’ -> ‘shippable/testresults/’
2017-02-06 10:13:58 ‘test/results/junit/.keep’ -> ‘shippable/testresults/.keep’
```
